### PR TITLE
build branding

### PR DESCRIPTION
**Pull request by ebarboni**
Hi,
 this is a side PR Proof of concept linked to discussion in #1270. Focusing on non-apidoc branding. 
  
It use git command to get hash and branches, build properties files to store them

`ReleaseJsonProperties.java` task is same as #1270 for apidoc + milestone support.
it basicaly put properties in files in build folder for later use by other ant build script.

Basic workflow for optionnal version that will impact title,splash,userdir,cache.
```
if hash is found for release111
   If vote is set, version number is the normal version 11.1 
   If no vote is set, 11.1-beta1
if hash is not found for release111
   11.1-dev will be used
```

json file is still in use from https://gitbox.apache.org/repos/asf?p=netbeans-jenkins-lib.git;a=blob_plain;f=meta/netbeansrelease.json

For testing in build.xml at line 121 it's possible to uncomment the echo to force a particular branch and hash.

### CODE REVIEW

**Comment by ebarboni** [comment]
Hi @neilcsmith-net and @JaroslavTulach , sorry I was very busy but does this change make some progress through the branding simplification ?

Regards.

**Comment by neilcsmith-net** [comment]
@ebarboni - no problem, me too until next week! In principle, this seems the right approach but I haven't tried it yet. How close to mergeable do you feel it is? Be good to merge and tidy up during this release cycle?

**Comment by geertjanw** [comment]
I like it and also the idea of merging this during a non-LTS release and therefore to not necessarily have it 100% done, but as a starting point for 12.0 LTS release. I.e., if it doesn't break anything and is basically OK, let's do it for 11.2.

File: nb/updatecenters/src/org/netbeans/modules/updatecenters/resources/Bundle.properties [hunk 1]
```diff
@@ -28,12 +28,13 @@ Services/AutoupdateType/pluginportal-update-provider.instance=NetBeans Plugin Po
 Services/AutoupdateType/82pluginportal-update-provider.instance=NetBeans 8.2 Plugin Portal
 
 #NOI18N
-URL_Distribution=https://netbeans.apache.org/nb/updates/dev/updates.xml.gz?{$netbeans.hash.code}
+URL_Distribution=@@metabuild.DistributionURL@@{$netbeans.hash.code}
```
[the remark this answers is not in the document]
Comment (neilcsmith-net) [reply nb/updatecenters/src/org/netbeans/modules/updatecenters/resources/Bundle.properties]: Changed as discussed via email.

**Comment by neilcsmith-net** [comment]
Well, probably still a little work-in-progress :smile: but merging into master for beta and we can tidy up as needed as we go.  Thanks!

**Comment by oyarzun** [comment]
I believe this change may have broke importing options since it tries to parse the build number.

https://issues.apache.org/jira/browse/NETBEANS-3198

**Comment by neilcsmith-net** [comment]
@oyarzun the bug report says it worked in the previous beta, so it's not directly this, but something in the build changes.  @ebarboni any thoughts?  Maybe follow up on issue - but from what I can see

This should be setting the `netbeans.buildnumber` property - https://github.com/apache/netbeans/blob/master/platform/o.n.bootstrap/src/org/netbeans/MainImpl.java#L265

But this is receiving a null - https://github.com/apache/netbeans/blob/master/platform/o.n.bootstrap/src/org/netbeans/MainImpl.java#L265

**Comment by oyarzun** [comment]
@neilcsmith-net I saw that it should be coming from MainImpl.java. Is it from the ide branding module?

The null is because it is expecting a date for the regex, e.g. 20190924, but we now use the git hash.

This is the manifest from Beta2

```
Manifest-Version: 1.0
Ant-Version: Apache Ant 1.10.6
Created-By: 1.8.0_191-b12 (Oracle Corporation)
OpenIDE-Module-Public-Packages: -
OpenIDE-Module-Java-Dependencies: Java > 1.6
OpenIDE-Module-Implementation-Version: 11.2-beta2-86e1b2eb194e8e9628a6
 6e6ee1a128134c70671a
AutoUpdate-Show-In-Client: false
OpenIDE-Module: org.netbeans.modules.ide.branding/1
OpenIDE-Module-Layer: org/netbeans/modules/ide/branding/layer.xml
OpenIDE-Module-Localizing-Bundle: org/netbeans/modules/ide/branding/Bu
 ndle.properties
OpenIDE-Module-Specification-Version: 1.38
OpenIDE-Module-Requires: org.openide.modules.ModuleFormat1
```

and 11.1

```
Manifest-Version: 1.0
Ant-Version: Apache Ant 1.9.7
Created-By: 1.8.0_131-b11 (Oracle Corporation)
OpenIDE-Module-Public-Packages: -
OpenIDE-Module-Java-Dependencies: Java > 1.6
OpenIDE-Module-Implementation-Version: netbeans-release-428-on-2019071
 6
AutoUpdate-Show-In-Client: false
OpenIDE-Module: org.netbeans.modules.ide.branding/1
OpenIDE-Module-Layer: org/netbeans/modules/ide/branding/layer.xml
OpenIDE-Module-Localizing-Bundle: org/netbeans/modules/ide/branding/Bu
 ndle.properties
OpenIDE-Module-Specification-Version: 1.37
OpenIDE-Module-Requires: org.openide.modules.ModuleFormat1
```
and 11.2 beta 1

```
Manifest-Version: 1.0
Ant-Version: Apache Ant 1.10.6
Created-By: 1.8.0_191-b12 (Oracle Corporation)
OpenIDE-Module-Public-Packages: -
OpenIDE-Module-Java-Dependencies: Java > 1.6
OpenIDE-Module-Implementation-Version: netbeans-TLP/netbeans/release11
 2-14-on-20190924
AutoUpdate-Show-In-Client: false
OpenIDE-Module: org.netbeans.modules.ide.branding/1
OpenIDE-Module-Layer: org/netbeans/modules/ide/branding/layer.xml
OpenIDE-Module-Localizing-Bundle: org/netbeans/modules/ide/branding/Bu
 ndle.properties
OpenIDE-Module-Specification-Version: 1.38
OpenIDE-Module-Requires: org.openide.modules.ModuleFormat1
```

**Comment by neilcsmith-net** [comment]
Fixed my links above. 

@oyarzun which regex?

**Comment by matthiasblaesing** [comment]
I debugged and the build number is correctly exported (not-null). The regexp @oyarzun references is this:

https://github.com/apache/netbeans/blob/master/platform/options.api/src/org/netbeans/modules/options/export/OptionsExportModel.java#L99

It is used here:

https://github.com/apache/netbeans/blob/master/platform/options.api/src/org/netbeans/modules/options/export/OptionsExportModel.java#L218

And the reason for the null value is here:

https://github.com/apache/netbeans/blob/master/platform/options.api/src/org/netbeans/modules/options/export/OptionsExportModel.java#L229

TL;DR: The module assumes, that the build number always contains a date pattern.

This is even tested:

https://github.com/apache/netbeans/blob/master/platform/options.api/test/unit/src/org/netbeans/modules/options/export/OptionsExportModelTest.java#L345
https://github.com/apache/netbeans/blob/master/platform/options.api/test/unit/src/org/netbeans/modules/options/export/OptionsExportModelTest.java#L362

**Comment by neilcsmith-net** [comment]
@matthiasblaesing thanks, yes just found the regex but not tests. Fun debugging via GitHub UI on a phone! :smile:

OK, so build number is never going to have a build date in it going forward! So, we'll need to decide a quick fix for 11.2 and a better one for 11.3.

**Comment by geertjanw** [comment]
Could we, as a quick fix, return a hardcoded date here, e.g., the date of beta3 for 11.2 beta3 and, then the date of the final release when we have it:

https://github.com/apache/netbeans/blob/master/platform/options.api/src/org/netbeans/modules/options/export/OptionsExportModel.java#L229

**Comment by neilcsmith-net** [comment]
@geertjanw yes, that's exactly what I've got in mind to do on Monday if no-one suggests anything different. I'm not quite sure at the moment why this code wants the build date in the first place, but that should at least work as a quick fix.

**Comment by matthiasblaesing** [comment]
For the "why is it there": 
https://github.com/apache/netbeans/blob/1b56603318f657219d54311be763cfba42fb8618/platform/options.api/src/org/netbeans/modules/options/export/OptionsChooserPanel.java#L439-L445

Looks like feature detection to me. Using a fixed date should make it work.

**Comment by neilcsmith-net** [comment]
Thanks @matthiasblaesing  Because build date makes so much more sense than IDE or module version there! :smile:  OK, hardcoding date fix now, and will do a search over uses of that system property as well.

And, JFTR, this was not caused by this PR at all - but #1540
